Fix authentication for new Blink API (OAuth2 - Nov 2025)

tpgal opened the pull request:
## Problem
The original script stopped working in mid-October 2024 due to Blink API changes.
The error "An app update is required" appeared on all login attempts.

## Solution
This PR updates the script to use the new OAuth2 authentication endpoint.

### Key Changes:
   - Migrated from API v4/v5 to OAuth2 endpoint (api.oauth.blink.com)
   - Added automatic token refresh mechanism
   - Added 2FA support via header

### Testing
   - Tested on Windows 11 with PowerShell 5.1+
   - Successfully authenticates with 2FA
   - Downloads videos and thumbnails correctly
   - Token refresh working after multiple hours of execution

Fixes issues reported in October-November 2025 by multiple users.

tpgal committed:
Fix authentication for new Blink API (OAuth2 - Nov 2025)

Fixes authentication issues caused by Blink API changes in October 2025.

Changes:
   - Migrated from API v4/v5 to OAuth2 endpoint (api.oauth.blink.com)
   - Added automatic token refresh mechanism
   - Added 2FA support via header

Tested and working as of November 2025.
Resolves the "An app update is required" error.

tpgal committed:
Updated to download thumbnails getting url from the homescreen endpoint

The homescreen endpoint is now used, providing all required info. The camera/usage endpoint is no longer needed.

tpgal commented:
Solved issue with thumbnails download, replacing the camera/usage endpoint by the homescreen endpoint, which provides the right full path to download the thumbnails.

will-k-2 commented:
@tpgal  Getting login failed as of today when trying to use are you getting the same error?

sutty550e commented:
By way of an extra datapoint, I too get login failed today, but not yesterday.

tpgal commented:
I'm sorry hearing that, unfortunately I'm on a business trip and won't be able to test it until the weekend. Hope Amazon hasn't decided to change it once again...

sutty550e commented:
No rush, I'm grateful for anything you can do, and whenever you can do it. Until 2 days ago, mine hadn't been working for a year, so it was a pleasant surprise to have it working at all. Odd that it only worked for a day, but I suppose they changed things again.

Just so you know, I did not get locked out from 'overuse', which I recall could be a thing, going back a good while in this project. I only logged in about three times, and only for a maximum of 10 minutes per time. Then, I tried again, a day later to check, after I got the update from @will-k-2 that he could not.

tpgal commented:
It looks like Amazon has done it again! I have the same issue, and we're not alone:
[https://github.com/fronzbot/blinkpy/issues/1147](https://github.com/fronzbot/blinkpy/issues/1147)
It will require futher analysis, let's hope it won't be as complex as last time and will just be a minor change...

tpgal committed:
New OAuth2 with PKCE Authorization Code Flow

Updated the login procedure with the new implemented PKCE authrization code flow.

tpgal commented:
I have updated once again the login procedure of the script, following the same approach as in blinkpy [#1149](https://github.com/fronzbot/blinkpy/pull/1149). Kudos to @KiraPC for his great work, and in record time!!
The script works again, and at least after the first login, you will receive an automatic email from The Blink Team informing you that a new device has been used to sign into your account (an iPhone 15 Pro! ;-), just ignore that message.

sutty550e commented:
Worked a treat for me. Thanks for your efforts, and also to @KiraPC.

tpgal commented:
@nayrk The script keeps working, and hopefully Amazon will not change the login procedure for a while. Could you please review this PR, and if it is fine, merge it into the main branch ?

KDN0691 commented:
Thanks to all, it ill run very fine. But what ist not clear for me, in my cloud are more than 500 videos, but only 109 are downloaded by the first request... My be the others will be loaded be the next time...

radiocricket commented:
@tpgal Your fixed worked well for so many months. I have to admit, I don't know when it stopped working since I fell behind on my downloads. When I just tried to download them today however, the login failed, but in an odd way. I did receive a text from Blink with my 2FA code. When I hit Enter, the window closes.
<img width="2136" height="1602" alt="image" src="https://github.com/user-attachments/assets/8019a63a-cd8a-47a5-b600-08f8dd3eb94f" />
The result is the same if I select Yes or All

sutty550e commented:
Exact same here. I didn't use it for a good while, but oddly noticed yesterday. Same failure messages, yet 2FA code is sent.

radiocricket commented:
I just checked and the last time I downloaded was June 5th, so roughly 2 weeks ago. I tried again today just hoping it was a hiccup but got the same failure. I did receive the verification code before selecting Y/A/N/L/S/?

tpgal commented:
I also noticed the issue a few days ago, this time it's a minor change by Amazon, they modified the return code of one of the APIs. I already have the patch and is working fine, I will upload it over the weekend.

sutty550e commented:
Fantastic! Well done, and thank you.

tpgal committed:
Add files via upload

Support to return code 202 for 2FA-signin

sutty550e commented:
I amended the 2FA flow, as you have shown, and my script is now working again perfectly. Thanks @tpgal

chrissv commented:
Confirmed working great - thanks!

Is it normal for the 2FA code to be requested each time you start the scripts? Is there any kind of "remember me" that would minimize the 2FA SMS messages (and the email notification of a "new device")?

radiocricket commented:
@tpgal Working great for me again too! Thank you so much for your time and willingness to help and share! Is there a way I can buy you a cup of coffee as a thank you?!

tpgal commented:
> @tpgal Working great for me again too! Thank you so much for your time and willingness to help and share! Is there a way I can buy you a cup of coffee as a thank you?!

It will be difficult, but thank you! 😄 Happy that the solution works and is useful for all of you!
